allocator_test.go

```diff
@@ -89,127 +100,166 @@ func TestCapacity(t *testing.T) {
+
+	clusterState := allocator.NewClusterState()
 	for i := 0; i < numNodes; i++ {
-		nodes.AddNode(int64(i), clusterCapacities[i])
+		clusterState.AddNode(
+			int64(i),
+			allocator.WithResourceOfNode(allocator.DiskResource, clusterCapacities[i]),
+		)
+	}
+	for i := 0; i < numReplicas; i++ {
+		clusterState.AddReplica(
+			int64(i),
+			rf,
+			allocator.WithDemandOfReplica(allocator.DiskResource, replicaSizeDemands[i]),
+		)
 	}
-	ranges := allocator.NewItems()
-	for i := 0; i < numRanges; i++ {
-		ranges.AddRange(int64(i), rf, rangeSizeDemands[i], 0)
-	}
-	status, allocation := allocator.New(allocator.WithResources()).Allocate(nodes, ranges)
+
+	status, allocation := allocator.Solve(clusterState, allocator.WithResources())
 	require.False(t, status)
 	require.Nil(t, allocation)
 }
 
 // Premise : test unhappy path and ensure we are not allocating when impossible to do so.
 func TestCapacityWithInsufficientNodes(t *testing.T) {
-	const numRanges = 10
+	const numReplicas = 10
 	const rf = 1
 	const numNodes = 3
 	clusterCapacities := []int64{70, 70, 70}
-	rangeSizeDemands := [numRanges]int64{85, 75, 12, 11, 10, 9, 8, 7, 6, 6}
-	nodes := allocator.NewCluster()
+	replicaSizeDemands := [numReplicas]int64{85, 75, 12, 11, 10, 9, 8, 7, 6, 6}
+
+	clusterState := allocator.NewClusterState()
 	for i := 0; i < numNodes; i++ {
-		nodes.AddNode(int64(i), clusterCapacities[i])
+		clusterState.AddNode(
+			int64(i),
+			allocator.WithResourceOfNode(allocator.DiskResource, clusterCapacities[i]),
+		)
+	}
+	for i := 0; i < numReplicas; i++ {
+		clusterState.AddReplica(
+			int64(i),
+			rf,
+			allocator.WithDemandOfReplica(allocator.DiskResource, replicaSizeDemands[i]),
+		)
 	}
-	ranges := allocator.NewItems()
-	for i := 0; i < numRanges; i++ {
-		ranges.AddRange(int64(i), rf, rangeSizeDemands[i], 0)
-	}
-	status, allocation := allocator.New(allocator.WithResources()).Allocate(nodes, ranges)
+
+	status, allocation := allocator.Solve(clusterState, allocator.WithResources())
 	require.False(t, status)
 	require.Nil(t, allocation)
 }
 
-// Premise : check tag affinity works on small cluster and range-set.
+// Premise : check tag affinity works on small cluster and replica-set.
 func TestTagsWithViableNodes(t *testing.T) {
-	const numRanges = 3
+	const numReplicas = 3
 	const rf = 1
 	const numNodes = 3
 	nodeTags := [][]string{
 		{"a=ant", "b=bus", "b=bin", "d=dog"},
 		{"a=all", "b=bus", "e=eat", "f=fun"},
 		{"a=art", "b=bin", "e=ear", "f=fur"},
 	}
-	rangeTags := [][]string{
+	replicaTags := [][]string{
 		{"a=art"},
 		{"e=eat"},
 		{"a=ant", "b=bus"},
 	}
-	nodes := allocator.NewCluster()
+
+	clusterState := allocator.NewClusterState()
 	for i := 0; i < numNodes; i++ {
-		nodes.AddNode(int64(i), 0, nodeTags[i]...)
-	}
-	ranges := allocator.NewItems()
-	for i := 0; i < numRanges; i++ {
-		ranges.AddRange(int64(i), rf, 0, 0, rangeTags[i]...)
+		clusterState.AddNode(
+			int64(i),
+			allocator.WithTagsOfNode(nodeTags[i]...),
+		)
```

Comment:
I can see the appeal of this approach, we wouldn't need to completely overhaul the API if we add additional `range/node` tags, although I do miss the simplicity of the old interface.